thrml/tempering.py

```diff
@@ -0,0 +1,188 @@
+        is_leaf=lambda a: isinstance(a, AbstractConditionalSampler),
+    )
+
+
+def _gibbs_steps(
+    key,
+    program: BlockSamplingProgram,
+    state_free: list,
+    state_clamp: list,
+    sampler_state: list,
+    n_iters: int,
+) -> tuple[list, list]:
+    """Run n_iters block-Gibbs sweeps for a single chain."""
+    if n_iters == 0:
+        return state_free, sampler_state
+
+    keys = jax.random.split(key, n_iters)
+    for k in keys:
+        state_free, sampler_state = sample_blocks(k, state_free, state_clamp, program, sampler_state)
+    return state_free, sampler_state
+
+
+def _attempt_swap_pair(
+    key,
+    ebm_i: AbstractEBM,
+    ebm_j: AbstractEBM,
+    program_i: BlockSamplingProgram,
+    program_j: BlockSamplingProgram,
+    state_i: list,
+    state_j: list,
+    clamp_state: list,
+):
+    """
+    Propose a swap between two adjacent temperature chains (i, j).
+
+    Acceptance ratio uses energies of both states under both models:
+    log r = (E_i(x_i) + E_j(x_j)) - (E_i(x_j) + E_j(x_i))
+    """
+    blocks_i = program_i.gibbs_spec.blocks
+    blocks_j = program_j.gibbs_spec.blocks
+
+    Ei_xi = ebm_i.energy(state_i + clamp_state, blocks_i)
+    Ej_xj = ebm_j.energy(state_j + clamp_state, blocks_j)
+    Ei_xj = ebm_i.energy(state_j + clamp_state, blocks_i)
+    Ej_xi = ebm_j.energy(state_i + clamp_state, blocks_j)
+
+    log_r = (Ei_xi + Ej_xj) - (Ei_xj + Ej_xi)
+    accept_prob = jnp.exp(jnp.minimum(0.0, log_r))
+    accept = jax.random.uniform(key) < accept_prob
+
+    if bool(accept):
+        return state_j, state_i, True
+    return state_i, state_j, False
+
+
+def _swap_pass(
+    key,
+    ebms: Sequence[AbstractEBM],
+    programs: Sequence[BlockSamplingProgram],
+    states: list[list],
+    sampler_states: list[list],
+    clamp_state: list,
+    offset: int,
+):
+    """Perform one swap pass over adjacent pairs with a given offset (0: (0,1),(2,3)...; 1: (1,2),(3,4)...)."""
+    n_pairs = len(ebms) - 1
+    accept_counts = [0] * n_pairs
+    attempt_counts = [0] * n_pairs
+
+    pair_indices = list(range(offset, n_pairs, 2))
+    if len(pair_indices) == 0:
+        return states, sampler_states, accept_counts, attempt_counts
+
+    keys = jax.random.split(key, len(pair_indices))
+    new_states = list(states)
+    new_sampler_states = list(sampler_states)
+
+    for idx, pair in enumerate(pair_indices):
+        i, j = pair, pair + 1
+        attempt_counts[pair] = 1
+        new_i, new_j, accepted = _attempt_swap_pair(
+            keys[idx], ebms[i], ebms[j], programs[i], programs[j], new_states[i], new_states[j], clamp_state
+        )
+        if accepted:
+            accept_counts[pair] = 1
+            new_states[i], new_states[j] = new_i, new_j
+            new_sampler_states[i], new_sampler_states[j] = new_sampler_states[j], new_sampler_states[i]
+
+    return new_states, new_sampler_states, accept_counts, attempt_counts
+
+
+def parallel_tempering(
+    key,
+    ebms: Sequence[AbstractEBM],
```

Comment:
Feels like we should have a high level wrapper that would just sample from the EBM and accept some beta type parameters